### ISSUE TITLE
float32 tl.sum result inconsistent with ATen eager reduction on Intel XPU

### ISSUE THREAD

**Issue by ZhaoqiongZ** [intel/intel-xpu-backend-for-triton#7197]
## Summary

`torch.compile` and eager produce different `float32` results for `x.sum(dim=-1)` on Intel GPU. The gap consistently exceeds
`torch.testing.assert_close` default `atol=1e-05`, causing a CI failure in PyTorch.

Only `torch.float32` is affected. `float16`, `bfloat16`, and `float64` agree between eager and compiled.

## Reproducer

```python
import torch

assert torch.xpu.is_available(), "XPU device required"

fn_compiled = torch.compile(lambda t: t.sum(dim=-1))
x0 = torch.randn(64, 4096, device="xpu", dtype=torch.float32)
_ = fn_compiled(x0)  # warm up

for seed in range(5):
    torch.manual_seed(seed)
    x = torch.randn(64, 4096, device="xpu", dtype=torch.float32)
    out_eager    = x.sum(dim=-1)
    out_compiled = fn_compiled(x)
    d = (out_eager.cpu() - out_compiled.cpu()).abs().max().item()
    status = "FAIL" if d > 1e-5 else "PASS"
    print(f"seed={seed}  eager vs compiled: {d:.3e}  {status}")
```

**Observed output (Intel Data Center GPU Max 1550):**

```
seed=0  eager vs compiled: 3.052e-05  FAIL
seed=1  eager vs compiled: 2.289e-05  FAIL
seed=2  eager vs compiled: 2.289e-05  FAIL
seed=3  eager vs compiled: 2.289e-05  FAIL
seed=4  eager vs compiled: 3.052e-05  FAIL
```

All five seeds fail consistently. The original CI shape `(64, 1024)` also fails but with a smaller (~7.6e-06 to ~1.1e-05) and config-dependent gap that makes it flaky. `(64, 4096)` always fails because it exercises the maximum persistent reduction size (see Analysis).

## Analysis

Shape `(64, 4096)` with `rnumel=4096` hits inductor's **persistent reduction** path: `rnumel * XBLOCK = 4096 * 1 ≤ MAX_PERSISTENT_BLOCK_NUMEL = 4096`. Only one autotuner config is generated (`XBLOCK=1`, `R0_BLOCK=4096`, `num_warps=8`):

```python
# inductor_meta: atomic_add_found=False
_tmp2 = tl.full([XBLOCK, R0_BLOCK], 0, tl.float32)
for r0_offset in tl.range(0, r0_numel, R0_BLOCK):   # runs once (persistent)
    tmp1 = load(...)
    _tmp2 = tl.where(r0_mask & xmask, _tmp2 + tmp1, _tmp2)
tmp2 = tl.sum(_tmp2, 1)   # <-- reduction over 4096 elements
tl.store(out_ptr0, tmp2)
```

All three computations — CPU sequential, XPU eager (ATen kernel), and XPU compiled (Triton `tl.sum`) — produce valid `float32` sums. They differ because float32 addition is not associative: different parallel reduction orderings yield slightly different results. For shape `(64, 4096)`, seed=0, worst-case row:

```
XPU eager    (ATen):   132.56774902   (err vs float64: 2.65e-05)
XPU compiled (Triton): 132.56777954   (err vs float64: 4.00e-06)
float64 reference:     132.56777554

eager vs compiled gap: 3.05e-05  >  atol=1e-05  -> FAIL
```

Notably, the Triton result is *closer* to the float64 reference than ATen. The issue is not which is more accurate, but that the two orderings diverge by more than `atol=1e-05`.

The gap scales with `rnumel` as `~sqrt(N) * eps * sigma`:

| shape | gap | status |
|---|---|---|
| (64, 256) | ~3.8e-06 | PASS |
| (64, 1024) | 7.6e-06 ~ 1.1e-05 | flaky (edge of atol) |
| (64, 2048) | ~1.5e-05 ~ 2.3e-05 | ALWAYS FAIL |
| (64, 4096) | ~1.9e-05 ~ 3.1e-05 | ALWAYS FAIL |

The kernel config for both shapes is `XBLOCK=1, num_warps=8` (256 threads, each thread accumulates 16 or 4 elements sequentially, then warp shuffle ×5 levels, then inter-warp SLM reduction ×3 levels). The ATen XPU reduction uses a different SYCL work-group decomposition, producing a different addition tree.

We are filing this for the Triton/XPU backend team to investigate whether `tl.sum` on XPU can be made numerically consistent with ATen eager.

## Impact

`test_combo_kernel_no_bench_persistent_reduction` in `pytorch/pytorch` is DISABLED on XPU due to the `(64, 1024)` shape failing:

- pytorch/pytorch issue: https://github.com/pytorch/pytorch/issues/187019
- torch-xpu-ops tracking: https://github.com/intel/torch-xpu-ops/issues/3973

## Environment

- **GPU**: Intel Data Center GPU Max 1550
- **PyTorch**: 2.13.0a0+git32364b4
- **oneAPI**: 2026.0

**Comment by wdziurdz** [comment]
@ZhaoqiongZ This isn't a Triton bug, it's just float32 non-associativity. I extended the reproducer to also compare against numpy on (64,4096):

``` python
import torch
import numpy as np

assert torch.xpu.is_available(), "XPU device required"

fn_compiled = torch.compile(lambda t: t.sum(dim=-1))
x0 = torch.randn(64, 4096, device="xpu", dtype=torch.float32)
_ = fn_compiled(x0)  # warm up

for seed in range(5):
    torch.manual_seed(seed)
    x = torch.randn(64, 4096, device="xpu", dtype=torch.float32)

    out_eager    = x.sum(dim=-1)
    out_compiled = fn_compiled(x)
    out_numpy    = torch.from_numpy(x.cpu().numpy().sum(axis=-1))  # numpy float32 sum

    # original comparison: eager (pytorch) vs compiled (triton)
    d_pt_tr = (out_eager.cpu() - out_compiled.cpu()).abs().max().item()
    # new: triton vs numpy
    d_tr_np = (out_compiled.cpu() - out_numpy).abs().max().item()
    # new: pytorch vs numpy
    d_pt_np = (out_eager.cpu() - out_numpy).abs().max().item()

    s = lambda d: "FAIL" if d > 1e-5 else "PASS"
    print(f"seed={seed}  "
          f"pytorch-vs-triton: {d_pt_tr:.3e} {s(d_pt_tr)}   "
          f"triton-vs-numpy: {d_tr_np:.3e} {s(d_tr_np)}   "
          f"pytorch-vs-numpy: {d_pt_np:.3e} {s(d_pt_np)}")

```

```
  seed=0  pytorch-vs-triton: 3.05e-05   triton-vs-numpy: 2.29e-05   pytorch-vs-numpy: 3.05e-05
  seed=1  pytorch-vs-triton: 2.29e-05   triton-vs-numpy: 3.05e-05   pytorch-vs-numpy: 2.29e-05
  seed=2  pytorch-vs-triton: 2.29e-05   triton-vs-numpy: 1.91e-05   pytorch-vs-numpy: 2.67e-05
  seed=3  pytorch-vs-triton: 1.53e-05   triton-vs-numpy: 3.05e-05   pytorch-vs-numpy: 3.05e-05
  seed=4  pytorch-vs-triton: 2.48e-05   triton-vs-numpy: 3.05e-05   pytorch-vs-numpy: 3.24e-05
```
Look at pytorch-vs-numpy: eager and numpy disagree by the same ~3e-5 with no Triton involved. So 1e-5 is unreachable for any two float32 sums of 4096 elements, not just Triton's.

I don't think we can make tl.sum match ATen bit-for-bit. Best fix is to bump the tolerance for float32 to ~3e-5, since 1e-5 is tighter than float32 can do for a sum this size.

@ZhaoqiongZ pls look on this.

**Comment by guangyey** [comment]
PyTorch does not use an `fp64` accumulator for `fp32` sum, and I believe Triton behaves the same way. That explains the relatively large numerical error and the corresponding tolerance. I agree with @wdziurdz that relaxing the tolerance is the right approach, consistent with https://github.com/pytorch/pytorch/pull/187825

**Comment by wdziurdz** [comment]
@guangyey @ZhaoqiongZ could we close issue ?
